File: java/Tries/InsertAndSearchWordsWithWildcards.java
import java.util.HashMap;
import java.util.Map;

public class TrieNode {
    Map<Character, TrieNode> children;
    boolean isWord;
    public TrieNode() {
        this.children = new HashMap<>();
        this.isWord = false;
    }
}

public class InsertAndSearchWordsWithWildcards {
    TrieNode root;

    public InsertAndSearchWordsWithWildcards() {
        this.root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode node = this.root;
        for (char c : word.toCharArray()) {
            node.children.putIfAbsent(c, new TrieNode());
            node = node.children.get(c);
        }
        node.isWord = true;
    }

    public boolean search(String word) {
        // Start searching from the root of the trie.
        return searchHelper(0, word, this.root);
    }

    private boolean searchHelper(int wordIndex, String word, TrieNode node) {
        for (int i = wordIndex; i < word.length(); i++) {
            char c = word.charAt(i);
            // If a wildcard character is encountered, recursively
            // search for the rest of the word from each child node.
            if (c == '.') {
                for (TrieNode child : node.children.values()) {
                    // If a match is found, return true.
                    if (searchHelper(i + 1, word, child)) {
                        return true;
                    }
                }
                return false;
            } else if (node.children.containsKey(c)) {
                node = node.children.get(c);
            } else {
                return false;
            }
        }
        // After processing the last character, return true if we've
        // reached the end of a word.
        return node.isWord;
    }
}

File: kotlin/Tries/InsertAndSearchWordsWithWildcards.kt
/*
    Definition of a TrieNode:

    data class TrieNode(
        var isWord: Boolean = false,
        val children: HashMap<Char, TrieNode> = hashMapOf()
    )
*/

class InsertAndSearchWordsWithWildcards {
    private val root = TrieNode()

    fun insert(word: String) {
        var node = root
        for (c in word) {
            if (c !in node.children) {
                node.children[c] = TrieNode()
            }
            node = node.children[c]!!
        }
        node.isWord = true
    }

    fun search(word: String): Boolean {
        // Start searching from the root of the trie.
        return searchHelper(0, word, root)
    }

    fun searchHelper(wordIndex: Int, word: String, node: TrieNode): Boolean {
        var node = node
        for (i in wordIndex until word.length) {
            val c = word[i]
            if (c == '.') {
                for (child in node.children.values) {
                    // If a match is found, return true.
                    if (searchHelper(i + 1, word, child)) {
                        return true
                    }
                }
            } else if (c in node.children) {
                node = node.children[c]!!
            } else {
                return false
            }
            // After processing the last character, return true if we've
            // reached the end of a word.
        }
        return node.isWord
    }
}
